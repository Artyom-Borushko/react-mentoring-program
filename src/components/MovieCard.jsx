import '../styles/movieCard.css';
import OptionsSelector from "./OptionsSelector";
import {movieDropdownOptions} from "../data/dropdownsOptions";

export default function MovieCard({title, releaseYear, relevantGenres, clickHandler, src}) {

    const handleDropdownSelection = (event) => {
        console.log('dropdown elements selected: ',
            event.target.innerText &&
            event.target.innerText.trim().toLowerCase()
        );
    };

    return (
        <div className={'movie-card'}>
            <OptionsSelector
                movieDropdownOptions={movieDropdownOptions}
                onDropdownSelection={handleDropdownSelection}
            />

            <img
                src={src}
                alt='movie card'
                aria-label={'movie card'}
                onClick={clickHandler}
            />
            <div className={'movie-name-year-wrapper'}>
                <p>{title}</p>
                <p className={'movie-release-year'}>{releaseYear}</p>
            </div>
            <p className={'movie-genres'}>{relevantGenres.join(', ')}</p>
        </div>
    );
}
