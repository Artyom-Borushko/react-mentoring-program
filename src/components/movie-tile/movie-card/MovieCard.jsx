import React, {memo} from "react";
import './movieCard.css';
import OptionsSelector from "../../shared/OptionsSelector";
import {movieDropdownOptions} from "../../../data/dropdownsOptions";
import {getMovieReleaseYear} from "../../../utilities/utilities";

export const MovieCard = memo(function MovieCard({movie, clickHandler, handleDropdownSelection}) {

    return (
        <div className={'movie-card'}>
            <OptionsSelector
                dropdownOptions={movieDropdownOptions}
                onDropdownSelection={(event) => handleDropdownSelection(event, movie)}
                controlImageSrc={'images/options-selector/threeDotsExpandOptionsIcon.png'}
                controlSource={'movie options'}
            />

            <img
                src={movie.src || '/images/movies-list/bohemianRhapsody.png'}
                alt='movie card'
                aria-label={'movie card'}
                onClick={() => clickHandler(movie)}
            />
            <div className={'movie-name-year-wrapper'}>
                <p>{movie.title}</p>
                <p className={'movie-release-year'}>{getMovieReleaseYear(movie.release_date)}</p>
            </div>
            <p className={'movie-genres'}>{movie.genres.join(', ')}</p>
        </div>
    );
})

export default MovieCard;
