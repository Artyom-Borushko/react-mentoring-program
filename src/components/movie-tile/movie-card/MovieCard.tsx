import React, { memo } from 'react';
import './movieCard.css';
import OptionsSelector from '../../shared/OptionsSelector';
import { movieDropdownOptions } from '../../../data/dropdownsOptions';
import { getMovieReleaseYear } from '../../../utilities/utilities';

interface Movie {
    src?: string;
    title: string;
    release_date: string;
    genres: string[];
}

interface ClickHandler {
    (movie: Movie): void;
}
interface HandleDropdownSelection {
    (event: React.SyntheticEvent, movie: Movie): void;
}

export const MovieCard = memo(({ movie, clickHandler, handleDropdownSelection }: {
    movie: Movie,
    clickHandler: ClickHandler,
    handleDropdownSelection: HandleDropdownSelection
}) => (
  <div className="movie-card">
    <OptionsSelector
      dropdownOptions={movieDropdownOptions}
      onDropdownSelection={(event) => handleDropdownSelection(event, movie)}
      controlImageSrc="images/options-selector/threeDotsExpandOptionsIcon.png"
      controlSource="movie options"
    />

    <img
      src={movie.src || '/images/movies-list/bohemianRhapsody.png'}
      alt="movie card"
      aria-label="movie card"
      onClick={() => clickHandler(movie)}
      role="presentation"
      data-testid="movie-card"
    />
    <div className="movie-name-year-wrapper">
      <p>{movie.title}</p>
      <p className="movie-release-year">{getMovieReleaseYear(movie.release_date)}</p>
    </div>
    <p className="movie-genres">{movie.genres.join(', ')}</p>
  </div>
));

export default MovieCard;
