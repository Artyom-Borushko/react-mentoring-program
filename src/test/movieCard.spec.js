import {render, screen} from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import '@testing-library/jest-dom';
import MovieCard from "../components/MovieCard";

describe("MovieCard component tests", () => {

    it("renders MovieCard component", () => {
        const {container} = render(<MovieCard
            title={'Kill1 Bill: Vol 1'}
            releaseYear={'1995'}
            relevantGenres={['comedy', 'crime', 'oscar']}
            src={'/images/movies-list/killBillVol2.png'}
        />);

        expect(container).toMatchSnapshot();
    });

    it("event listener is working clicking on image", async () => {
        const mockClickHandler = jest.fn();
        render(<MovieCard
                title={'Kill1 Bill: Vol 1'}
                releaseYear={'1995'}
                relevantGenres={['comedy', 'crime', 'oscar']}
                src={'/images/movies-list/killBillVol2.png'}
                clickHandler={mockClickHandler}
            />);

        await userEvent.click(screen.getByRole('img', {name: 'movie card'}));

        expect(mockClickHandler).toHaveBeenCalledTimes(1);
    });

    it("event listener is not working clicking on movie releaseYear", async () => {
        const mockClickHandler = jest.fn();
        render(<MovieCard
            title={'Kill1 Bill: Vol 1'}
            releaseYear={'1995'}
            relevantGenres={['comedy', 'crime', 'oscar']}
            src={'/images/movies-list/killBillVol2.png'}
            clickHandler={mockClickHandler}
        />);

        await userEvent.click(screen.getByText('1995'));

        expect(mockClickHandler).toHaveBeenCalledTimes(0);
    });
});
