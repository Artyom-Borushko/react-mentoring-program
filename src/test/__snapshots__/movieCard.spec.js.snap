// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard component tests renders MovieCard component 1`] = `
<div>
  <div
    class="movie-card"
  >
    <div>
      <img
        alt="expand options"
        aria-label="expand options"
        class="movie-expand-options-icon"
        src="images/options-selector/threeDotsExpandOptionsIcon.png"
      />
    </div>
    <img
      alt="movie card"
      aria-label="movie card"
      src="/images/movies-list/killBillVol2.png"
    />
    <div
      class="movie-name-year-wrapper"
    >
      <p>
        Kill1 Bill: Vol 1
      </p>
      <p
        class="movie-release-year"
      >
        1995
      </p>
    </div>
    <p
      class="movie-genres"
    >
      comedy, crime, oscar
    </p>
  </div>
</div>
`;
